### PR TITLE
4/Fixing Lag and Integrating DialogueBoxService with Items

### LINKED ISSUE #161: Fix Lag Caused from Interaction of Friendly NPC and Player
# **Description**
**Task:** [Fix Chat Overlay Lag](161).
**Feature:** #152

From its implementation in sprint one, the chat box overlay does not function as well as intended. This was due to the lag caused when a player is in close proximity to a friendly NPC. This was cause the game to create a new chat overlay each time to be used and then disposed of afterwards. This ticket will solve this problem. Additionally items had the same issues hence their implementation will also be refactored to make the service.
 
## **Example (optional)**
- [x] Fix lag caused by unnecessary creation and deletion of chat overlay each time.
 
# **Milestones**
List of steps that need to be completed for this task.
- [x] Remove use of dispose() (Sep. 9)
- [x] Implement a method to hide chat overlay to be called on and used on a later interaction with the NPC (Sep. 9)
 
# **Documentation**
- TBD.
 
# **Member**
- Bailey Stoodley @BaileyS03

### PR DESCRIPTION
**Pull request by BaileyS03**
Fixes lag when being around items & when talking to the entity it loads on main game screen reducing lag caused later in the game. 

Please ensure there are no extra issues as I did have difficulties rebasing of main which was my own but please keep this in mind :)

Closes #161

### CODE REVIEW

File: source/core/src/main/com/csse3200/game/areas/ForestGameArea.java
Comment (callumreillyy) [inline source/core/src/main/com/csse3200/game/areas/ForestGameArea.java:1]: Why was the Quest Popup() removed? Assuming it has been loaded elsewhere.

**Review by callumreillyy — APPROVED** [review]
All changes made to factory classes look fine. Only query is the removal of QuestPopup(), but I assume this has been added elsewhere.

File: source/core/src/main/com/csse3200/game/components/tasks/ItemProximityTask.java [hunk 1]
```diff
@@ -20,7 +21,7 @@ public class ItemProximityTask extends ProximityTask {
     private static final Logger logger = LoggerFactory.getLogger(ItemProximityTask.class);
     private final AbstractItem item;
     private boolean itemPickedUp = false;
-    DialogueBox itemOverlay;
+    private boolean hasApproached;
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/components/tasks/ItemProximityTask.java:24]: This should be inherited as a protected variable from ProximityTask

File: source/core/src/main/com/csse3200/game/components/tasks/ItemProximityTask.java [hunk 2]
```diff
@@ -95,15 +82,21 @@ public int getPriority() {
...
-                itemOverlay = null;
-            }
+        ServiceLocator.getEntityChatService().hideCurrentOverlay();
+    }
+
+    /**
+     * Checks if the player is near the item using the distance between the item and the player
+     * @return true if the player is within the proximity threshold, false otherwise.
+     */
+    private boolean isPlayerNearItem() {
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/components/tasks/ItemProximityTask.java:92]: This method is unused and it's functionality is provided by ProximityTask

**Review by mitsudz — CHANGES_REQUESTED** [review]
Other than the 2 things I mentioned looks pretty good. I think the merge of item proximity task and proximity task just went slightly wrong and some old code was kept.

File: source/core/src/main/com/csse3200/game/components/tasks/ItemProximityTask.java (hunk 2 above)
Comment (BaileyS03) [reply source/core/src/main/com/csse3200/game/components/tasks/ItemProximityTask.java]: Fixed

File: source/core/src/main/com/csse3200/game/components/tasks/ItemProximityTask.java (hunk 1 above)
Comment (BaileyS03) [reply source/core/src/main/com/csse3200/game/components/tasks/ItemProximityTask.java]: Fixed

**Review by mitsudz — APPROVED** [review]

**Review by LucasRyall — APPROVED** [review]
Looks good, just wondering if the friendly npcs need the combat stats in their config files. If they do need them, move them to be declared in the base entity and remove them from both friendly and enemy cause no point declaring it in 2 places.

**Pull request merged**